relayer/message_coordinator.go

```diff
@@ -0,0 +1,254 @@
+	sourceBlockchainID ids.ID,
+	originSenderAddress common.Address,
+	destinationBlockchainID ids.ID,
+	destinationAddress common.Address,
+) *ApplicationRelayer {
+	// Check for an exact match
+	applicationRelayerID := database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		originSenderAddress,
+		destinationAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+
+	// Check for a match on sourceBlockchainID and destinationBlockchainID, with a specific originSenderAddress and any destinationAddress
+	applicationRelayerID = database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		originSenderAddress,
+		database.AllAllowedAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+
+	// Check for a match on sourceBlockchainID and destinationBlockchainID, with any originSenderAddress and a specific destinationAddress
+	applicationRelayerID = database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		database.AllAllowedAddress,
+		destinationAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+
+	// Check for a match on sourceBlockchainID and destinationBlockchainID, with any originSenderAddress and any destinationAddress
+	applicationRelayerID = database.CalculateRelayerID(
+		sourceBlockchainID,
+		destinationBlockchainID,
+		database.AllAllowedAddress,
+		database.AllAllowedAddress,
+	)
+	if applicationRelayer, ok := mc.applicationRelayers[applicationRelayerID]; ok {
+		return applicationRelayer
+	}
+	mc.logger.Debug(
+		"Application relayer not found. Skipping message relay.",
+		zap.String("blockchainID", sourceBlockchainID.String()),
+		zap.String("destinationBlockchainID", destinationBlockchainID.String()),
+		zap.String("originSenderAddress", originSenderAddress.String()),
+		zap.String("destinationAddress", destinationAddress.String()),
+	)
+	return nil
+}
+
+func (mc *MessageCoordinator) ProcessManualWarpMessage(
+	warpMessage *relayerTypes.WarpMessageInfo,
+) (common.Hash, error) {
+	// Send any messages that were specified in the configuration
+	mc.logger.Info(
+		"Relaying manual Warp message",
+		zap.String("blockchainID", warpMessage.UnsignedMessage.SourceChainID.String()),
+		zap.String("warpMessageID", warpMessage.UnsignedMessage.ID().String()),
+	)
+	appRelayer, handler, err := mc.GetAppRelayerMessageHandler(warpMessage)
+	if err != nil {
+		mc.logger.Error(
+			"Failed to parse manual Warp message.",
+			zap.Error(err),
+			zap.String("warpMessageID", warpMessage.UnsignedMessage.ID().String()),
+		)
+		return common.Hash{}, err
+	}
+
+	return appRelayer.ProcessMessage(handler)
+}
+
+func (mc *MessageCoordinator) ProcessMessage(blockchainID ids.ID, messageID common.Hash, blockNum *big.Int) (common.Hash, error) {
```

Comment:
I'd suggest renaming this to `ProcessMessageID` to distinguish it from `ApplicationRelayer.ProcessMessage`, which operates on the full message, and doesn't fetch it from the chain.